add ansible playbook runner to ansiblegate module

gtmanfred opened the pull request:
### What does this PR do?
Adds the ability to run ansible playbooks from salt execution modules or from states, specifically for orchestration states.

### Tests written?

Added

### Commits signed with GPG?

Yes

gtmanfred committed:
add ansible playbook runner to ansiblegate module

terminalmage commented on salt/states/ansiblegate.py:
```diff
@@ -99,3 +103,85 @@ def __virtual__():
+    if ansible_kwargs is None:
+        ansible_kwargs = {}
+    checks = __salt__['ansible.playbooks'](name, rundir=rundir, check=True, diff=True, **ansible_kwargs)
+    if all(not check['changed'] for _, check in six.iteritems(checks['stats'])):
```
You can use `six.itervalues()` here and not need to use the "throwaway" variable.

gtmanfred replied:
cool

terminalmage reviewed (changes requested):
Just one minor change I would make, otherwise this is ++.

gtmanfred committed:
use itervalues

gtmanfred commented:
Still need to write tests for this.

isbm commented on salt/modules/ansiblegate.py:
```diff
@@ -276,3 +277,99 @@ def list(pattern=None):
+        salt 'ansiblehost'  ansible.playbook playbook=/srv/playbooks/play.yml
+    '''
+    command = ['ansible-playbook', playbook]
+    if check is True:
```
What is the reason to explicitly check the type alongside the value? I.e. why `if check:` is not sufficient?

isbm commented on salt/modules/ansiblegate.py:
```diff
@@ -276,3 +277,99 @@ def list(pattern=None):
+    if diff is True:
+        command.append('--diff')
+    if extra_vars is not None:
+        command.append('--extra-vars={0}'.format(json.dumps(extra_vars)))
```
Hmm, still the same question though. I'd say `if extra_vars:`

isbm commented on salt/states/ansiblegate.py:
```diff
@@ -99,3 +103,85 @@ def __virtual__():
+    return salt.fileclient.get_file_client(__opts__)
+
+
+def _changes(plays):
```
Docstring?

isbm reviewed (changes requested):
Generally, if we want to validate the params, we should do it explicitly. Otherwise I would use Pythonic checks (see the comment).

Otherwise, excellent, woohoo!! :sunglasses:

gtmanfred committed:
appease BOFH

Also, handle @ for extra_vars

gtmanfred committed:
add ansible playbook test

gtmanfred commented:
Test added

gtmanfred commented:
re-run py

isbm commented on salt/states/ansiblegate.py:
```diff
@@ -99,3 +103,88 @@ def __virtual__():
+        'comment': 'Running playbook {0}'.format(name),
+        'name': name,
+    }
+    if gitrepo is not None:
```
If `gitrepo` is an empty string this will fail.

isbm commented on salt/states/ansiblegate.py:
```diff
@@ -99,3 +103,88 @@ def __virtual__():
+            **git_kwargs,
+        )
+    if ansible_kwargs is None:
+        ansible_kwargs = {}
```
Maybe this is better way, unless we do not want to validate all that and crash with the error "You've passed an empty string instead of a parameters" (or something like that):

```python
if not ansible_kwargs:
    ansible_kwargs = {}
```

gtmanfred replied:
I am just going to check if it a dict and if it isn't then i will set it to {}.

isbm reviewed (changes requested):
Cool, thanks! Some last little nitpicks though.

rallytime commented:
Lint: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/22834/violations/

gtmanfred committed:
fix pylint

handle git_kwargs and ansible_kwargs better as well.

gtmanfred committed:
fix pylint and test

gtmanfred commented:
This should be all set to go on the next test run.

gtmanfred commented:
https://jenkins.saltstack.com/job/PR/job/salt-pr-rs-cent7-n/20023/testReport/integration.states.test_ansiblegate/AnsiblePlaybooksTestCase/

terminalmage reviewed (commented):
Is `gitrepo` part of a naming convention used elsewhere? If not, it may make sense to change it. I'm just worried about confusion between `gitrepo` (no underscore) and `git_kwargs` (underscore). It seems like this could lead to confusion/typos.

isbm reviewed (changes requested):
@terminalmage actually a _very_ good catch, it slipped out of my radar: the `git_<something>` is better convention than `git<something>`.

@gtmanfred could you please appease BOFH (and terminalmage) by looking into name conventions and checking to `None` one more time? :wink: I would appreciate that.

Other then that, it is good to go as of my current views.

gtmanfred committed:
These darn BOFHs <3

isbm reviewed (approved):
LOL :laughing:  :+1: good to go.

rallytime commented:
@gtmanfred Lint :)

gtmanfred committed:
fix pylint

isbm commented on salt/states/ansiblegate.py:
```diff
@@ -99,3 +105,94 @@ def __virtual__():
+        run nginx install:
+          ansible.playbooks:
+            - name: install.yml
+            - gitrepo: git://github.com/gituser/playbook.git
```
Docs :)

gtmanfred replied:
bah... i did a search replace... apparently i am just terrible at regex.

isbm commented on tests/integration/states/test_ansiblegate.py:
```diff
@@ -0,0 +1,73 @@
+        ret = self.run_state(
+            'ansible.playbooks',
+            name='install.yml',
+            gitrepo='git://github.com/gtmanfred/playbooks.git',
```
There is no such param as `gitrepo` anymore (and so this should be fixed). But why the test is passing anyway, makes me wondering if the `run_state` is reliable enough reporting wrong parameter has been passed?

gtmanfred replied:
It did fail this time https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/23939/testReport/junit/integration.states.test_ansiblegate/AnsiblePlaybooksTestCase/test_ansible_playbook/

With not having gitrepo in the keys.

isbm reviewed (changes requested):
@gtmanfred docs, tests. :)

gtmanfred committed:
aaahhhh more gitrepo -> git_repo

gtmanfred committed:
add to fluorine release notes

isbm reviewed (approved):
@gtmanfred thanks! Hopefully now it is OK to go.

rallytime commented:
@gtmanfred You've got a test failure here: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/24000/

gtmanfred committed:
gitrepo -> git_repo in test

gtmanfred commented:
re-run py

rallytime commented:
@gtmanfred The Py3 tests are failing your new test. Can you take a look?

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-cent7-py3/6163/

- integration.states.test_ansiblegate.AnsiblePlaybooksTestCase.test_ansible_playbook

gtmanfred commented:
Yup, i am working on it now.

It is failing because ansible can't be installed on py3.4, so i was installing it with only py2 in salt-jenkins.

But this makes it so the ansiblegate modules don't load because the module stuff stops if it can't import ansible, which this doesn't need.

gtmanfred committed:
Load ansible.playbooks if binary is installed

Don't load extra ansible module stuff is ansible is not importable.

Use @depends decorator https://docs.saltstack.com/en/latest/ref/modules/#depends-decorator

gtmanfred commented:
Tests should be fixed on this run.

gtmanfred committed:
fix for pylint

gtmanfred committed:
fix unittest

Ansiblegate should always be included now, the list and help functions
will just be removed in the loader when Depends.enforce_dependencies is
run.

gtmanfred commented:
Ok, this should be ready.

rallytime commented:
@gtmanfred There is still a test failing here, unfortunately:

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/24138/testReport/junit/unit.loader.test_globals/LoaderGlobalsTest/test_states/

gtmanfred committed:
handle wrapped functions for loader globals test

gtmanfred commented:
re-run py

gtmanfred committed:
handle wrapped on py2 as well

gtmanfred commented:
re-run all

rallytime commented:
re-run py

rallytime commented:
@gtmanfred I think the following test is related. I am not seeing this fail elsewhere:

- integration.modules.test_cmdmod.CMDModuleTest.test_run

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/11261/

gtmanfred committed:
fix cmdmod test

gtmanfred commented:
🎉
